Documentation for 5.10 features (#3818)
plotly/plotly.py: doc/python/selections.md

    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.13.7
  kernelspec:
    display_name: Python 3 (ipykernel)
    language: python
    name: python3
  language_info:
    codemirror_mode:
      name: ipython
      version: 3
    file_extension: .py
    mimetype: text/x-python
    name: python
    nbconvert_exporter: python
    pygments_lexer: ipython3
    version: 3.9.0
  plotly:
    description: How to use selections in Python. Examples of adding and styling selections.
    display_as: file_settings
    language: python
    layout: base
    name: Selections
    order: 26
    permalink: python/selections/
    thumbnail: thumbnail/make_selection.jpg
---

## Adding Selections

*New in 5.10*

You can add persistent selections to a rendered figure using the **Box Select** and **Lasso Select** tools in the mode bar.
To add multiple selections, select **Shift** when making new selections.
To clear a selection, double-click it. You can clear all selections by double-clicking any unselected area of the figure.

You can also add selections to a figure that displays when it renders using `fig.add_selection`.

> Here, we add a rectangular selection that connects the points `3.0`, `6.5`, `3.5`, and `5.5`.

Reviewer: A point here should be described with x and y positions. So perhaps: 
```suggestion
Here, we add a rectangular selection with a region between `3.0` and `6.5` on the x axis and between `3.5` and `5.5` on the y axis.
```

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(type="rect", x0=3.0, y0=6.5, x1=3.5, y1=5.5)

fig.show()
```

> 

Reviewer: The `path` option should be described here as well.

Author: Thanks @archmoj. Trying to add this now. Should this render something?
It is the same as the earlier example, but just uses path.
Nothing shows on the figure with this.

```
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(type="path", path="M2000,65L4000,75L8000,65Z")

fig.show()

```

Reviewer: It should. What happens if you draw a `shape` instead?
```py
fig.add_shape(type="path", path="M2000,65L4000,75L8000,65Z")
```

Author: Yeah, this one works. 
So actually it seems like the selection is only immediately visible in the range of the existing data.

So, this works:

```

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(type="path", path="M2,6L4,7L4,6Z")
#fig.add_selection(type="path", path="M20,6L4000,7L8,65Z")

fig.show()
```

<img width="690" alt="image" src="https://user-images.githubusercontent.com/12749831/180281338-55ebdfe4-c44c-4fd5-b5a6-eb81535f3108.png">

Added an example now.

## Styling Selections

In the above example, we added a selection to the figure that is displayed when the figure renders.
`fig.add_selection` accepts additional properties that you can use to style the selection. Here, we add a `color`, `width`, and specify the `dash` type for the selection.

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(type="rect",
    x0=2.5, y0=6.5, x1=3.5, y1=5.5,
    line=dict(
        color="Crimson",
        width=2,
        dash="dash",
    ))

fig.show()

```

## Fill Color for Active Selections

> You can style active selections with `activeselection`. In this example, we set active selections to appear with a `fillcolor` of `blue`.

Reviewer: ```suggestion
You can style the active selection with `activeselection`. In this example, we set active selections (while created or when clicked) to appear with a `fillcolor` of `yellow`.
```

Author: Updated

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(type="rect", x0=3.0, y0=6.5, x1=3.5, y1=5.5)

fig.update_layout(dragmode='select',

>                   activeselection=dict(fillcolor='blue'))

Reviewer: The `blue` may be too dark.
```suggestion
                  activeselection=dict(fillcolor='yellow'))
```

      name: ipython
      version: 3
    file_extension: .py
    mimetype: text/x-python
    name: python
    nbconvert_exporter: python
    pygments_lexer: ipython3
    version: 3.9.0
  plotly:
    description: How to use selections in Python. Examples of adding and styling selections.
    display_as: file_settings
    language: python
    layout: base
    name: Selections
    order: 26
    permalink: python/selections/
    thumbnail: thumbnail/make_selection.jpg
---

## Adding Selections

*New in 5.10*

You can add persistent selections to a rendered figure using the **Box Select** and **Lasso Select** tools in the mode bar.
To add multiple selections, select **Shift** when making new selections.
To clear a selection, double-click it. On a subplot you can clear all selections by double-clicking any unselected area of the subplot.

You can also add selections to a figure that displays when it renders using `fig.add_selection`.
Here, we add a rectangular selection with a region between `3.0` and `6.5` on the x axis and between `3.5` and `5.5` on the y axis.

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")

> fig.add_selection(type="rect", x0=3.0, y0=6.5, x1=3.5, y1=5.5)

Reviewer: I think we could drop the type here too. Or we may decide to keep it if that's helpful information.
```
fig.add_selection(x0=3.0, y0=6.5, x1=3.5, y1=5.5)
```

Author: Added. Good point. Didn't realize it wasn't required.

fig.show()
```

## Selections using Path

> In the above example, we added a rectangular selection (`type="rect"`). You can also render a custom SVG for a selection by setting `type="path"` and defining a `path`. Here, we create a selection with the path "M2,6.5L4,7.5L4,6Z"

Reviewer: ```suggestion
In the above example, we added a rectangular selection (`type="rect"`). You can also render a custom SVG for a selection by defining a `path` that can include single or multiple polygons. Here, we create a selection with a single polygon path "M2,6.5L4,7.5L4,6Z". 

Please note that multiple polygons e.g. "M0,0L0,10L10,10,L10,0Z M2,2L2,8L8,8,L8,2Z" could be used to subtract certain regions from the selection.
```

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")

> fig.add_selection(type="path", path="M2,6.5L4,7.5L4,6Z")

Reviewer: I think we could omit the type here.
```suggestion
fig.add_selection(path="M2,6.5L4,7.5L4,6Z")
```

fig.show()

> ```

Reviewer: Let's also describe `xref` and `yref` then provide an example with multiple subplots please. Something similar to https://github.com/plotly/plotly.js/commit/b183c9570c4bf5f3b48859004832dd3f65f60555.

Author: @archmoj , I just added an example for this. Do I have the general idea right?

> 

Reviewer: I suggest we also provide an example using `date` axis to illustrate how to position selection regions.
Something like https://github.com/plotly/plotly.js/commit/1107600f7a9be2c11cd31c800bd46675d1545acf

Author: I've added an example.

---
jupyter:
  jupytext:
    notebook_metadata_filter: all
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.13.7
  kernelspec:
    display_name: Python 3 (ipykernel)
    language: python
    name: python3
  language_info:
    codemirror_mode:
      name: ipython
      version: 3
    file_extension: .py
    mimetype: text/x-python
    name: python
    nbconvert_exporter: python
    pygments_lexer: ipython3
    version: 3.9.0
  plotly:
    description: How to use selections in Python. Examples of adding and styling selections.
    display_as: file_settings
    language: python
    layout: base
    name: Selections
    order: 26
    permalink: python/selections/
    thumbnail: thumbnail/make_selection.jpg
---

> ## Adding Selections

Reviewer: ```suggestion
## Adding Selections to Cartesian subplots
```

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(x0=3.0, y0=6.5, x1=3.5, y1=5.5)

fig.update_layout(dragmode='select',
                  activeselection=dict(fillcolor='yellow'))

fig.show()
```

## Styling New Selections

You can style new selections made on the figure by setting properties on `newselection`.
Try making a new selection on the figure to try it out.

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")

fig.update_layout(dragmode='select',
                  newselection=dict(line=dict(color='blue')))

fig.show()
```

## Selections with Time Series

Selections are also supported on time series figures. Here, we add a rectangular selection with a region between the dates `2019-01-01"` and `"2019-10-01"` on the x axis and between `0.95` and `1.15` on the y axis.

```python
import plotly.express as px

df = px.data.stocks()
fig = px.line(df, x='date', y="GOOG")
fig.add_selection(x0="2019-01-01", y0=0.95, x1="2019-10-01", y1=1.15)
fig.show()
```

> ## Selections on Subplots

Reviewer: ```suggestion
## Selections on Multiple Cartesian Subplots
```

Reviewer: Or perhaps better
```suggestion
## Referencing Selections on Multiple Cartesian Subplots
```

> You can add selections to subplots by specifying axis ids using `xaxis` and `yaxis`. Here, we add one selection on the plot with axis ids `x` and `y2` and two selections to the the plot with axis ids `x` and `y`.

Reviewer: ```suggestion
You can add selections to multiple Cartesian subplots by specifying `xref` and/or `yref`. Here, we add one selection on the plot with axis ids `x` and `y2` and two selections to the the plot with axis ids `x` and `y`.
```

Reviewer: Please also note that one don't have to provide `x` and `y` values while the default values are `x` and `y`.

```python
import plotly.graph_objects as go

import numpy as np

t = np.linspace(-1, 1.2, 2000)

> x = (t**3) + (0.3 * np.random.randn(2000))

Reviewer: Did you provide a seed for these random numbers?
`The random seed should be set if using randomly-generated data in new/modified examples.`

Author: I have now. Thanks

Reviewer: Great. Then please check the box in the PR description too :)

    ),
    yaxis = dict(
        zeroline = False,
        domain = [0,0.85],
        showgrid = False
    ),
    xaxis2 = dict(
        zeroline = False,
        domain = [0.85,1],
        showgrid = False
    ),
    yaxis2 = dict(
        zeroline = False,
        domain = [0.85,1],
        showgrid = False
    ),
    height = 600,
    width = 600,
    bargap = 0,
    hovermode = 'closest',
    showlegend = False, 
    selections = [
        dict(
            x0 = 0.5,
            x1 = -0.5,
            xref = "x",
            y0 = 190,
            y1= 0,
            yref = "y2",
            line = dict(
                color="yellow"
            )
        ),
        dict(
            x0 = -0.2,
            x1 = -1.5,
            xref = "x",
            y0 = 2,
            y1= -1,
            yref = "y",
            line = dict(
                color="yellow"
            )
        ),   
        dict(
            path= "M0.75,2.39L0.98,3.38L1.46,3.68L1.80,3.35L2.01,2.51L1.67,1.15L1.18,0.50L0.65,0.66L0.54,0.83L0.49,1.56Z",
            xref= 'x',
            yref = 'y',
            line = dict(
                color='yellow'
            )
        )
    ]
)

fig.show()
```

> 

Reviewer: I may be a good idea to provide an example here for `splom` traces as well.
In this example https://github.com/plotly/plotly.js/commit/6eba38d08bdc55f54905e44058896d50c51dcc62 multiple points are selected on different `splom` subplots.

Author: Thanks @archmoj, I've added an example based on yours

---
jupyter:
  jupytext:
    notebook_metadata_filter: all
    text_representation:
      extension: .md
      format_name: markdown
      format_version: '1.3'
      jupytext_version: 1.13.7
  kernelspec:
    display_name: Python 3 (ipykernel)
    language: python
    name: python3
  language_info:
    codemirror_mode:
      name: ipython
      version: 3
    file_extension: .py
    mimetype: text/x-python
    name: python
    nbconvert_exporter: python
    pygments_lexer: ipython3
    version: 3.9.0
  plotly:
    description: How to use selections in Python. Examples of adding and styling selections.
    display_as: file_settings
    language: python
    layout: base
    name: Selections

>     order: 26

Author: ```suggestion
    order: 38
```

    permalink: python/selections/
    thumbnail: thumbnail/make_selection.jpg
---

> ## Adding Selections to Cartesian Subplots

Reviewer 2: Overall this document is great, but one thing which is missing is some indication to the reader of how to retrieve the selection values! We will need a link to some Dash documentation or something that folks can follow to actually do something with the selection once the user does it!

Do the selected values come back out using `fig.full_figure_for_development()` in a way that we can statically demo in the docs here?

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(path="M2,6.5L4,7.5L4,6Z")

fig.show()
```

## Styling Selections

In the above example, we added a selection to the figure that is displayed when the figure renders.
`fig.add_selection` accepts additional properties that you can use to style the selection. Here, we add a `color`, `width`, and specify the `dash` type for the selection.

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(
    x0=2.5, y0=6.5, x1=3.5, y1=5.5,
    line=dict(
        color="Crimson",
        width=2,
        dash="dash",
    ))

fig.show()

```

## Fill Color for Active Selections

You can style the active selection with `activeselection`. In this example, we set active selections (when created or clicked) to appear with a `fillcolor` of `yellow`.

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")
fig.add_selection(x0=3.0, y0=6.5, x1=3.5, y1=5.5)

fig.update_layout(dragmode='select',
                  activeselection=dict(fillcolor='yellow'))

fig.show()
```

> ## Styling New Selections

Reviewer: Let's move this section above `Fill Color for Active Selections` section as it is a more useful attribute to know.

You can style new selections made on the figure by setting properties on `newselection`.
Try making a new selection on the figure to try it out.

```python
import plotly.express as px

df = px.data.iris()

fig = px.scatter(df, x="sepal_width", y="sepal_length")

fig.update_layout(dragmode='select',
                  newselection=dict(line=dict(color='blue')))

fig.show()
```

## Selections with Time Series

Selections are also supported on time series figures. Here, we add a rectangular selection with a region between the dates `2019-01-01"` and `"2019-10-01"` on the x axis and between `0.95` and `1.15` on the y axis.

```python
import plotly.express as px

df = px.data.stocks()
fig = px.line(df, x='date', y="GOOG")
fig.add_selection(x0="2019-01-01", y0=0.95, x1="2019-10-01", y1=1.15)
fig.show()
```

> ## Referencing Selections on a Scatterplot Matrix

Reviewer: Let's move this section below `Referencing Selections on Multiple Cartesian Subplots` section as it is more common use case. Thanks!

fig.show()
```

## Selections with Time Series

Selections are also supported on time series figures. Here, we add a rectangular selection with a region between the dates `2019-01-01"` and `"2019-10-01"` on the x axis and between `1` and `1.15` on the y axis.

```python
import plotly.express as px

df = px.data.stocks()
fig = px.line(df, x='date', y="GOOG")

> fig.add_selection(x0="2019-01-01", y0=1, x1="2019-10-01", y1=1.17)

Reviewer: ```suggestion
fig.add_selection(x0="2019-01-01", y0=1, x1="2019-10-01", y1=1.15)
```
